python/lsst/source/injection/inject_engine.py

```diff
@@ -0,0 +1,462 @@
+        else:
+            shear_data.update({shear_param: source_data[shear_param]})
+    try:
+        object = object.shear(**shear_data)
+    except TypeError:
+        pass
+    # Apply the instrumental flux and return.
+    object = object.withFlux(inst_flux)
+    return object
+
+
+def make_galsim_trail(
+    source_data: dict[str, Any],
+    wcs: SkyWcs,
+    sky_coords: SpherePoint,
+    inst_flux: float,
+    trail_thickness: float = 1e-6,
+) -> galsim.transform.Transformation:
+    """Make a trail with GalSim from a collection of source data.
+
+    Parameters
+    ----------
+    source_data : `dict` [`str`, `Any`]
+        Dictionary of source data.
+    wcs : `lsst.afw.geom.SkyWcs`
+        World coordinate system.
+    sky_coords : `lsst.geom.SpherePoint`
+        Sky coordinates of the source.
+    inst_flux : `float`
+        Instrumental flux of the source.
+    trail_thickness : `float`
+        Thickness of the trail in pixels.
+
+    Returns
+    -------
+    object : `galsim.transform.Transformation`
+        A fully specified and transformed GalSim object.
+    """
+    # Make a 'thin' box to mimic a line surface brightness profile of default
+    # thickness = 1e-6 (i.e., much thinner than a pixel)
+    object = galsim.Box(source_data["trail_length"], trail_thickness)
+    try:
+        object = object.rotate(source_data["beta"] * galsim.degrees)
+    except KeyError:
+        pass
+    object = object.withFlux(inst_flux * source_data["trail_length"])  # type: ignore
+    # GalSim objects are assumed to be in sky-coords. As we want the trail to
+    # appear as defined above in image-coords, we must transform it here.
+    linear_wcs = wcs.linearizePixelToSky(sky_coords, arcseconds)
+    mat = linear_wcs.getMatrix()
+    object = object.transform(mat[0, 0], mat[0, 1], mat[1, 0], mat[1, 1])
+    return object  # type: ignore
+
+
+def make_galsim_stamp(
+    stamp_file: str,
+    fits_alignment: str,
+    wcs: SkyWcs,
+    sky_coords: SpherePoint,
+    inst_flux: float,
+) -> galsim.transform.Transformation:
+    """Make a postage stamp with GalSim from a FITS file.
+
+    Parameters
+    ----------
+    stamp_file : `str`
+        Path to the FITS file containing the postage stamp.
+    fits_alignment : `str`
+        Alignment of the FITS image to the WCS. Allowed values: "wcs", "pixel".
+    wcs : `lsst.afw.geom.SkyWcs`
+        World coordinate system.
+    sky_coords : `lsst.geom.SpherePoint`
+        Sky coordinates of the source.
+    inst_flux : `float`
+        Instrumental flux of the source.
+
+    Returns
+    -------
+    object: `galsim.transform.Transformation`
+        A fully specified and transformed GalSim object.
+    """
+    stamp_file = stamp_file.strip()
+    if os.path.exists(stamp_file):
+        # Loop over each HDU (max 100) until a viable ImageHDU has been found.
+        stamp_data = False
+        hdu_counter = 0
+        while hdu_counter <= 100 and not stamp_data:
```

Comment:
`100` feels really arbitrary to me.  What's the use case here?

Reply:
It is somewhat arbitrary. It's not beyond the scope of imagination that the first 10 HDUs may be filled with tabular data, but 1000 seemed too large. An order of magnitude argument led me to 100. Happy to modify as you see fit.

Reply:
On discussion, we'll remove this `while` loop and replace it with a `for` loop with an upper limit determined by astropy's FITS reader.

Reply:
Using `astropy.io.fits`, no `for` loop is actually required at all here. I've replaced this whole block with this:

```
    if os.path.exists(stamp_file):
        with fits.open(stamp_file) as hdul:
            hdu_images = [hdu.is_image and hdu.size > 0 for hdu in hdul]  # type: ignore
        if any(hdu_images):
            stamp_data = galsim.fits.read(stamp_file, read_header=True, hdu=np.where(hdu_images)[0][0])
        else:
            raise RuntimeError(f"Cannot find image in input FITS file {stamp_file}.")
```

That seems to nicely cover all bases; thanks for catching this!